src/azure_filesystem.cpp

```diff
@@ -0,0 +1,436 @@
+			if (to_read > new_buffer_available) {
+				ReadRange(hfh, location + buffer_offset, (char *)buffer + buffer_offset, to_read);
+				hfh.buffer_available = 0;
+				hfh.buffer_idx = 0;
+				hfh.file_offset += to_read;
+				break;
+			} else {
+				ReadRange(hfh, hfh.file_offset, (char *)hfh.read_buffer.get(), new_buffer_available);
+				hfh.buffer_available = new_buffer_available;
+				hfh.buffer_idx = 0;
+				hfh.buffer_start = hfh.file_offset;
+				hfh.buffer_end = hfh.buffer_start + new_buffer_available;
+			}
+		}
+	}
+}
+
+bool AzureStorageFileSystem::FileExists(const string &filename) {
+	try {
+		auto handle = OpenFile(filename, FileFlags::FILE_FLAGS_READ);
+		auto &sfh = handle->Cast<AzureStorageFileHandle>();
+		if (sfh.length == 0) {
+			return false;
+		}
+		return true;
+	} catch (...) {
+		return false;
+	};
+}
+
+void AzureStorageFileSystem::ReadRange(FileHandle &handle, idx_t file_offset, char *buffer_out, idx_t buffer_out_len) {
+	auto &afh = handle.Cast<AzureStorageFileHandle>();
+
+	try {
+		// Specify the range
+		Azure::Core::Http::HttpRange range;
+		range.Offset = (int64_t)file_offset;
+		range.Length = buffer_out_len;
+		Azure::Storage::Blobs::DownloadBlobToOptions options;
+		options.Range = range;
+		options.TransferOptions.Concurrency = afh.read_options.transfer_concurrency;
+		options.TransferOptions.InitialChunkSize = afh.read_options.transfer_chunk_size;
+		options.TransferOptions.ChunkSize = afh.read_options.transfer_chunk_size;
+		auto res = afh.blob_client.DownloadTo((uint8_t *)buffer_out, buffer_out_len, options);
+
+	} catch (Azure::Storage::StorageException &e) {
+		throw IOException("AzureStorageFileSystem Read to " + afh.path + " failed with " + e.ErrorCode +
+		                  "Reason Phrase: " + e.ReasonPhrase);
+	}
+}
+
+AzureParsedUrl AzureStorageFileSystem::ParseUrl(const string &url) {
+	string container, prefix, path;
+
+	if (url.rfind("azure://", 0) * url.rfind("az://", 0) != 0) {
+		throw IOException("URL needs to start with azure:// or az://");
+	}
+	auto prefix_end_pos = url.find("//") + 2;
+	auto slash_pos = url.find('/', prefix_end_pos);
+	if (slash_pos == string::npos) {
+		throw IOException("URL needs to contain a '/' after the host");
+	}
+	container = url.substr(prefix_end_pos, slash_pos - prefix_end_pos);
+	if (container.empty()) {
+		throw IOException("URL needs to contain a bucket name");
+	}
+
+	prefix = url.substr(0, prefix_end_pos);
+	path = url.substr(slash_pos + 1);
+	return {container, prefix, path};
+}
+
+std::shared_ptr<AzureContextState> AzureStorageFileSystem::GetOrCreateStorageAccountContext(FileOpener *opener,
+                                                                                            const std::string &path) {
+	auto *client_context = FileOpener::TryGetClientContext(opener);
+
+	AzureContextState *result = nullptr;
+
+	auto &storage_account = client_context->registered_state[DEFAULT_AZURE_STORAGE_ACCOUNT];
+	try {
```

Comment:
Yes I wanted to avoid the two lookup, but you are right it an unordered map so it should be almost like O(1) complexity. I kept some old habit when only map was available is the std^^
Done in [ec242c0](https://github.com/duckdb/duckdb_azure/pull/32/commits/ec242c057fda27b125314901b3e45d4e4db4189c)